### PR TITLE
Implement ADviaFDConfig

### PR DESCRIPTION
**Pull request by mzgubic**
Closes https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/241 and https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/221

### CODE REVIEW

File: src/rule_config.jl
Comment (oxinabox) [inline src/rule_config.jl:19]: why the `@kwdef` ?

File: src/rule_config.jl
Comment (oxinabox) [inline src/rule_config.jl:28]: Why is this not
```suggestion
    primals = (f, args...)
```
?

File: src/rule_config.jl
Comment (oxinabox) [inline src/rule_config.jl:44]: Do we need to fill in `is_ignored`?
Or would it be ok to return invalid derivatives for these ones.
I suspect the answer is that it would not be OK.
But I wish it wasn't.

I think this section should have a comment explaining that the reason we are calling `auto_primal_and_tangent` is just to try and workout which things we should set `is_ignored` for.

**Review by oxinabox — COMMENTED** [review]
should this try and use a `rrule`/`frule` first then only fallback to using the FiniteDifferences if that returns `nothing`?

File: src/rule_config.jl
Comment (mzgubic) [reply src/rule_config.jl:44]: I think we do need it, because otherwise we get dimension mismatch in `_make_j′vp_call`

Added a comment that explains the block

**Comment by mzgubic** [comment]
>should this try and use a rrule/frule first then only fallback to using the FiniteDifferences if that returns nothing?

Do we ever want to avoid using the rule and make sure to use FD for the inner rule? I don't think so.

I think this means we can just combine the `ADviaRuleConfig` and `ADviaFDConfig` into a `TestConfig`, which does what you describe? And it's not a breaking change because it's not exported?

**Comment by oxinabox** [comment]
> I think this means we can just combine the ADviaRuleConfig and ADviaFDConfig into a TestConfig, which does what you describe?

Yeah, I guess so.

> And it's not a breaking change because it's not exported?

It is, because people use it directly. Don't they?
But we can `@deprecate_binding ADviaRuleConfig TestConfig false`

File: src/rule_config.jl
Comment (oxinabox) [inline src/rule_config.jl:44]: Shall we avoid creating a variable that is never used again?
The following fuses
```suggestion
    is_ignored = isa.(tangent.(primals_and_tangents), NoTangent)
```

**Review by oxinabox — APPROVED** [review]
merge when happy

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:96]:
```suggestion
    test_frule(TestConfig(), args...; kwargs...)
```

File: test/testers.jl
Comment (oxinabox) [inline test/testers.jl:288]: lol

**Review by oxinabox — APPROVED** [review]

File: test/rule_config.jl
Comment (mzgubic) [inline test/rule_config.jl:106]: @oxinabox FYI there was actually a gross bug in the original implementation.

This `TestConfig` will be used to test rules for higher order functions `f(inner, array)` where we don't care that much for the rule for `inner` and want to do finite differencing for it out of convenience.

However, if we _don't check whether a rule exists_ inside `f/rrule_via_ad` using the `TestConfig`, we are actually ignoring the outer rule implementation (for `f`) entirely, and always pass the tests because finite differencing is done on the outer rule 😅 

I will wait for you to reply before merging since this was a significant development (even if not a significant change to the code)

Comment (oxinabox) [reply test/rule_config.jl:106]: oh

**Pull request merged**
